### ISSUE TITLE
Compile errors in the editor

### ISSUE THREAD

**Issue by jrieken** [Microsoft/vscode#16819]
Since today (Insiders-1.8.0 0d65ced0fa7398dc9d50c39a23a2a8cc8406ff18) I see compile errors in the editor that don't show up when running `gulp watch`. Unsure if this is new TypeScript behaviour or bugs? Samples are

* https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostConfiguration.ts#L110
* https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/snippet/common/snippetParser.ts#L208

**Comment by kieferrm** [comment]
This is the condensed version of `snippetParser.ts` that still creates the problem:
```typescript
export abstract class Marker {
	static toString(markers: Marker[]): string {
		return '';
	}
}

export class Placeholder extends Marker {
	constructor(public name: string = '', public defaultValue: Marker[]) {
		super();
	}
}

function foo(marker: Marker) {
	if (marker instanceof Placeholder) {
		console.log(marker.name);
		console.log(marker.defaultValue);
	}
}

foo(null);
```

**Comment by kieferrm** [comment]
This is the error we are getting:
![screen shot 2016-12-07 at 8 55 21 am](https://cloud.githubusercontent.com/assets/4674940/20977609/f931ae68-bc5a-11e6-8537-560ffa32ceef.png)

**Comment by jrieken** [comment]
Thanks for drilling into this. I might push a change to rename my method...

**Comment by kieferrm** [comment]
Here is the second one:
```ts
export interface Foo {
	property: boolean;
	[key: string]: any;
}

function freezer(bar: Foo): Foo {
	return Object.freeze(bar);
}

freezer(null);
```

The issue is caused when an index is combined with anything else no matter if a property or method.

The error we get is:
![screen shot 2016-12-07 at 9 36 11 am](https://cloud.githubusercontent.com/assets/4674940/20979185/a9d68680-bc60-11e6-85d5-d792c5a9926e.png)

**Comment by mjbvz** [comment]
Opened two more issues against TS to track issue we are seeing:

* https://github.com/Microsoft/TypeScript/issues/12725
* https://github.com/Microsoft/TypeScript/issues/12729


I'm going to submit quick workarounds fixes to our code for those two problems. They just require additional casts

**Comment by mjbvz** [comment]
Fixed those two errors with #16861

**Comment by mjbvz** [comment]
I believe this was resolved by #16861 and 499a648.

We also followed up with the TS team. Most of the errors are by design and are breaking changes in TS 2.1
